### ISSUE TITLE
PHP 5.4 using PHP 5.5 features

### ISSUE THREAD

**Issue by assertchris** [icicleio/icicle#3]
Hey friend,

https://github.com/icicleio/Icicle/blob/5797e370acb24e64c70494a35f4aad0f9de325c3/src/Coroutine/Coroutine.php#L132-L134

`finally` is a PHP 5.5 feature, but `phpunit.xml.dist` and `composer.json` both stipulate PHP 5.4 as a minimum version. Am I missing something or should the minimum version be bumped to 5.5?

Edit: also, generator syntax is [`(PHP 5 >= 5.5.0)`](http://php.net/manual/en/language.generators.overview.php)

**Comment by trowski** [comment]
The Coroutine component requires PHP 5.5, but everything else only requires 5.4. I did this so the potential user base would be as large as possible. (Honestly I wish I could make it 5.6 only for variadic functions and eliminate some of the `array_slice()` and `func_get_args()` non-sense that's otherwise needed.) The _requirements_ section of `README.md` points out that corountines need PHP 5.5, but maybe that needs to be more obvious.

I'm developing some additional components (DNS, HTTP, Process, etc) that will only require 5.4 as well.

With those things in mind, do you have any thoughts on where the version requirement should be? I'd like to keep the user base as large as possible, but I also don't want there to be confusion.

**Comment by assertchris** [comment]
Ok, I understand. That makes sense. I think it would be better if the 5.4 stuff and the 5.5 stuff were in different repositories though.

**Comment by trowski** [comment]
That would make sense, but I think I would still talk about Coroutines on the main page of the primary Icicle repository, since I feel like it's one of the main reasons to use Icicle.

Would this separation make sense?

EventEmitter - This component could be separated so anyone could use it in any project, since it doesn't depend on the event loop at all.
Icicle - (Loop, Promise, Stream, Socket) - These are the core components of Icicle. I could separate all of these into separate repositories as well... but would it make sense to separate everything?
Coroutine - Because it requires PHP 5.5.

**Comment by trowski** [comment]
After some consideration, I've decided to just raise the minimum PHP version required for Icicle to 5.5.

The entire point of the library is to make writing asynchronous code easier with coroutines, and I realized I didn't want to limit additional components to only 5.4, requiring them to only use only promises rather than both promises and coroutines.

While I may limit the audience in the short term, I think this is a better decision for the library as a whole moving forward. PHP 5.4 has already entered security support only... it doesn't make sense to limit the library because of a version that will be unsupported in 6 months. Using coroutines will allow for more rapid development of additional components.
